### PR TITLE
Fix p5-p6 account nonce and sequence number migration bug 

### PR DESCRIPTION
**Pull request by EmilLa1**
## Purpose

Fix a bug where account nonces and seqnumbers where not properly migrated over in p5->p6

Fixes: https://github.com/Concordium/concordium-node/issues/905

## Changes

_Describe the changes that were needed.

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [x] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

File: concordium-consensus/src/Concordium/KonsensusV1/TreeState/Implementation.hs [hunk 1]
```diff
@@ -301,9 +309,14 @@ mkInitialSkovData ::
     Duration ->
     -- |Bakers at the genesis block
     EpochBakers ->
+    -- |If present the initial skov data created
+    -- will carry over the account nonces and sequence numbers
+    -- from the provided transaction table.
+    -- This is used for migration.
+    Maybe TransactionTable ->
```
Comment (abizjak) [inline concordium-consensus/src/Concordium/KonsensusV1/TreeState/Implementation.hs:316]: I'd recommend not complicating things with an optional here, and just require it here, and pass an emptyTransctionTable in a couple of places in tests.

File: concordium-consensus/src/Concordium/KonsensusV1/TreeState/Implementation.hs [hunk 2]
```diff
@@ -301,9 +309,14 @@ mkInitialSkovData ::
...
     -- |The initial 'SkovData'
     SkovData pv
-mkInitialSkovData rp genMeta genState _currentTimeout _skovEpochBakers =
+mkInitialSkovData' rp genMeta genState _currentTimeout _skovEpochBakers maybeTT =
```
Comment (abizjak) [inline concordium-consensus/src/Concordium/KonsensusV1/TreeState/Implementation.hs:319]:
```suggestion
mkInitialSkovData rp genMeta genState _currentTimeout _skovEpochBakers transactionTable =
```

File: concordium-consensus/src/Concordium/KonsensusV1/TreeState/Implementation.hs [hunk 3]
```diff
@@ -322,7 +335,7 @@ mkInitialSkovData rp genMeta genState _currentTimeout _skovEpochBakers =
                 }
         _persistentRoundStatus = initialPersistentRoundStatus
         _roundStatus = initialRoundStatus _currentTimeout genesisBlockPointer
-        _transactionTable = TT.emptyTransactionTable
+        _transactionTable = fromMaybe TT.emptyTransactionTable maybeTT
```
Comment (abizjak) [inline concordium-consensus/src/Concordium/KonsensusV1/TreeState/Implementation.hs:338]:
```suggestion
        _transactionTable = transactionTable
```

File: concordium-consensus/src/Concordium/KonsensusV1/TreeState/Implementation.hs (hunk 1 above)
Comment (abizjak) [reply concordium-consensus/src/Concordium/KonsensusV1/TreeState/Implementation.hs:316]: That will also simplify `initialiseNewSkovV1` where you don't have to override the transaction table after constructing the state anymore, i.e., no longer need to

```
                    mkInitialSkovData gscRuntimeParameters genMeta pbs genTimeoutDuration genEpochBakers
                        & transactionTable .~ genTT
```

File: concordium-consensus/src/Concordium/KonsensusV1/TreeState/Implementation.hs [hunk 4]
```diff
@@ -343,6 +356,51 @@ mkInitialSkovData rp genMeta genState _currentTimeout _skovEpochBakers =
...
+    GenesisMetadata ->
+    -- |Genesis statep
+    PBS.HashedPersistentBlockState pv ->
+    -- |The base timeout
+    Duration ->
+    -- |Bakers at the genesis block
+    EpochBakers ->
+    -- |The initial 'SkovData'
+    SkovData pv
+mkInitialSkovData rp genMeta genState _currentTimeout _skovEpochBakers = mkInitialSkovData' rp genMeta genState _currentTimeout _skovEpochBakers Nothing
```
Comment (abizjak) [inline concordium-consensus/src/Concordium/KonsensusV1/TreeState/Implementation.hs:375]:
```suggestion
```

File: concordium-consensus/src/Concordium/KonsensusV1/TreeState/Implementation.hs [hunk 5]
```diff
@@ -343,6 +356,51 @@ mkInitialSkovData rp genMeta genState _currentTimeout _skovEpochBakers =
...
+    -- |The transaction table to migrate from.
+    -- This carries over the account nonces and sequence numbers to the
+    -- new transaction table.
+    TransactionTable ->
+    -- |The initial 'SkovData'
+    SkovData pv
+mkInitialSkovDataWithTT rp genMeta genState _currentTimeout _skovEpochBakers tt =
+    let migratedTT = emptyTransactionTableWithSequenceNumbers migratedAccountNonces migratedUpdateSequenceNumbers
+        migratedAccountNonces = map (\(addr, anft) -> (aaeAddress addr, anft ^. anftNextNonce)) (HM.toList $ tt ^. ttNonFinalizedTransactions)
+        migratedUpdateSequenceNumbers = Map.fromList $ map (\(utype, nfcu) -> (utype, nfcu ^. nfcuNextSequenceNumber)) (Map.toList $ tt ^. ttNonFinalizedChainUpdates)
```
Comment (abizjak) [inline concordium-consensus/src/Concordium/KonsensusV1/TreeState/Implementation.hs:401]: I don't think this is correct. When this function is used you pass in the existing transaction table, with potentially pending transactions.

And then you only extract these sequence numbers but don't retain the transaction themselves. So unless I'm missing something this won't really work the way it's meant to, transactions will be dropped.

Is there a reason for doing it in a different way than we already do for existing protocol versions, where we just copy over the transaction table?

File: concordium-consensus/src/Concordium/KonsensusV1/TreeState/Implementation.hs (hunk 1 above)
Comment (EmilLa1) [reply concordium-consensus/src/Concordium/KonsensusV1/TreeState/Implementation.hs]: Moving over the TT and PT and cleaned up as suggested.

File: concordium-consensus/src/Concordium/KonsensusV1/TreeState/Implementation.hs [hunk 6]
```diff
@@ -290,6 +291,9 @@ purgeRoundExistingQCs rnd = roundExistingQCs %=! snd . Map.split (rnd - 1)
 -- |Create an initial 'SkovData pv'
 -- This constructs a 'SkovData pv' from a genesis block
 -- which is suitable to grow the tree from.
+--
+-- In the case that this is from a genesis, then an empty transaction table
+-- and
```
Comment (abizjak) [inline concordium-consensus/src/Concordium/KonsensusV1/TreeState/Implementation.hs:296]: and?

File: concordium-consensus/src/Concordium/KonsensusV1/TreeState/Implementation.hs [hunk 7]
```diff
@@ -301,9 +305,13 @@ mkInitialSkovData ::
     Duration ->
     -- |Bakers at the genesis block
     EpochBakers ->
+    -- |'TransactionTable' to initialize the 'SkovData' with.
+    TransactionTable ->
+    -- |'PendingTransactionTable' to initialize the 'SkovData' with.
```
Comment (abizjak) [inline concordium-consensus/src/Concordium/KonsensusV1/TreeState/Implementation.hs:310]: Please document here like we do for the V0 consensus, what this pending table must satisfy. This is quite tricky and easy to miss.

**Review by abizjak — APPROVED** [review]

File: concordium-consensus/src/Concordium/GlobalState/Persistent/TreeState.hs [hunk 8]
```diff
@@ -1014,7 +1014,19 @@ instance
...
+                    -- part of a block.
+                    -- This ensures that the transaction gets purged after the protocol
+                    -- update if has expired.
+                    -- If we carried over the @CommitPoint@ then the transaction will
+                    -- not get purged, and the only way it could get evicted from the
+                    -- 'TransactionTable' is, if it becomes part of a new block that gets
+                    -- finalized.
+                    -- However this may never be the case if the transaction has become
+                    -- deprecated as part of the protocol update.
+                    Committed{..} -> (bi, Received{_tsCommitPoint = 0, ..})
```
Comment (abizjak) [inline concordium-consensus/src/Concordium/GlobalState/Persistent/TreeState.hs:1029]: Shouldn't you then do the same in Implementation.hs for that clearOnProtocolUpdate?

Comment (EmilLa1) [reply concordium-consensus/src/Concordium/GlobalState/Persistent/TreeState.hs:1029]: Yes.

File: concordium-consensus/src/Concordium/GlobalState/Persistent/TreeState.hs [hunk 9]
```diff
@@ -1014,7 +1014,19 @@ instance
...
                     _ -> (bi, s)
```
Comment (td202) [inline concordium-consensus/src/Concordium/GlobalState/Persistent/TreeState.hs:1030]: This is not sufficient. We need to set the commit point of `Received` transactions to 0 also.

File: concordium-consensus/src/Concordium/GlobalState/Persistent/TreeState.hs [hunk 10]
```diff
@@ -1014,7 +1014,19 @@ instance
...
+                    -- Mark all @Committed@ transactions as @Received@.
+                    -- Note that we here set the @CommitPoint@ to "0", effectively
+                    -- treating the transaction as just received and it has never been
+                    -- part of a block.
+                    -- This ensures that the transaction gets purged after the protocol
+                    -- update if has expired.
+                    -- If we carried over the @CommitPoint@ then the transaction will
+                    -- not get purged, and the only way it could get evicted from the
+                    -- 'TransactionTable' is, if it becomes part of a new block that gets
+                    -- finalized.
```
Comment (td202) [inline concordium-consensus/src/Concordium/GlobalState/Persistent/TreeState.hs:1026]: This is not strictly the only way -- the commit point could also pass. The point is that the old commit point doesn't make any sense in a renewed consensus.

File: concordium-consensus/src/Concordium/KonsensusV1/TreeState/Implementation.hs [hunk 11]
```diff
@@ -290,6 +291,17 @@ purgeRoundExistingQCs rnd = roundExistingQCs %=! snd . Map.split (rnd - 1)
...
+-- In the case that this is from a genesis, then an empty transaction table
+-- and empty pending transaction table must be provided.
+--
+-- Invariants:
+-- In case that a non-empty transaction table is supplied, then we're migrating
+-- from an existing state, thus the 'TransactionTable' and 'PendingTransactionTable'
+-- passed in must both be from the former state.
+--  * If there are any transactions in the 'TransactionTable' then the corresponding 'PendingTransactionsTable'
+--    must take these transactions into account. Refer to 'PendingTransactionTable' for more details.
+--  * The caller must make sure, that the supplied  'TransactionTable' does NOT contain any @Committed@ transactions.
```
Comment (td202) [inline concordium-consensus/src/Concordium/KonsensusV1/TreeState/Implementation.hs:304]: Also the commit point of `Received` transactions should be 0.

File: concordium-consensus/src/Concordium/KonsensusV1/TreeState/Implementation.hs [hunk 12]
```diff
@@ -1080,7 +1096,19 @@ clearOnProtocolUpdate = do
...
+                -- This ensures that the transaction gets purged after the protocol
+                -- update if has expired.
+                -- If we carried over the @CommitPoint@ then the transaction will
+                -- not get purged, and the only way it could get evicted from the
+                -- 'TransactionTable' is, if it becomes part of a new block that gets
+                -- finalized.
+                -- However this may never be the case if the transaction has become
+                -- deprecated as part of the protocol update.
+                (bi, TT.Committed{..}) -> (bi, TT.Received{TT._tsCommitPoint = 0, ..})
                 s -> s
```
Comment (td202) [inline concordium-consensus/src/Concordium/KonsensusV1/TreeState/Implementation.hs:1112]: As before - we need to set the commit point to 0 on ALL transactions.

**Review by td202 — CHANGES_REQUESTED** [review]

File: concordium-consensus/src/Concordium/KonsensusV1/TreeState/Implementation.hs [hunk 13]
```diff
@@ -301,7 +301,8 @@ purgeRoundExistingQCs rnd = roundExistingQCs %=! snd . Map.split (rnd - 1)
 -- passed in must both be from the former state.
 --  * If there are any transactions in the 'TransactionTable' then the corresponding 'PendingTransactionsTable'
 --    must take these transactions into account. Refer to 'PendingTransactionTable' for more details.
---  * The caller must make sure, that the supplied  'TransactionTable' does NOT contain any @Committed@ transactions.
+--  * The caller must make sure, that the supplied 'TransactionTable' does NOT contain any @Committed@ transactions
+--    and all transactions either 'TT.Received' or 'TT.Committed' must have their 'TT.CommitPoint' set to 0.
```
Comment (td202) [inline concordium-consensus/src/Concordium/KonsensusV1/TreeState/Implementation.hs:305]:
```suggestion
--    and all transactions have their commit point set to 0.
```
There cannot be committed transactions, so we don't have to talk about them.

**Review by td202 — APPROVED** [review]

**Pull request merged**